modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/LocalDatabases.java

```diff
@@ -0,0 +1,208 @@
+
+    private final GeoIpCache cache;
+    private final Path geoipConfigDir;
+
+    private final Map<String, DatabaseReaderLazyLoader> defaultDatabases;
+    private volatile Map<String, DatabaseReaderLazyLoader> configDatabases;
+
+    LocalDatabases(Environment environment, GeoIpCache cache) {
+        this(
+            // In GeoIpProcessorNonIngestNodeTests, ingest-geoip is loaded on the classpath.
+            // This means that the plugin is never unbundled into a directory where the database files would live.
+            // Therefore, we have to copy these database files ourselves. To do this, we need the ability to specify where
+            // those database files would go. We do this by adding a plugin that registers ingest.geoip.database_path as an
+            // actual setting. Otherwise, in production code, this setting is not registered and the database path is not configurable.
+            environment.settings().get("ingest.geoip.database_path") != null ?
+                getGeoipConfigDirectory(environment) :
+                environment.modulesFile().resolve("ingest-geoip"),
+            environment.configFile().resolve("ingest-geoip"),
+            cache);
+    }
+
+    LocalDatabases(Path geoipModuleDir, Path geoipConfigDir, GeoIpCache cache) {
+        this.cache = cache;
+        this.geoipConfigDir = geoipConfigDir;
+        this.configDatabases = Map.of();
+        this.defaultDatabases = initDefaultDatabases(geoipModuleDir);
+    }
+
+    void initialize(ResourceWatcherService resourceWatcher) throws IOException {
+        configDatabases = initConfigDatabases(geoipConfigDir);
+
+        FileWatcher watcher = new FileWatcher(geoipConfigDir);
+        watcher.addListener(new GeoipDirectoryListener());
+        resourceWatcher.add(watcher, ResourceWatcherService.Frequency.HIGH);
+
+        LOGGER.info("initialized default databases [{}], config databases [{}] and watching [{}] for changes",
+            defaultDatabases.keySet(), configDatabases.keySet(), geoipConfigDir);
+    }
+
+    DatabaseReaderLazyLoader getDatabase(String name) {
+        DatabaseReaderLazyLoader database = configDatabases.get(name);
+        if (database != null) {
+            return database;
+        } else {
+            return defaultDatabases.get(name);
+        }
+    }
+
+    List<DatabaseReaderLazyLoader> getAllDatabases() {
+        List<DatabaseReaderLazyLoader> all = new ArrayList<>(defaultDatabases.values());
+        all.addAll(configDatabases.values());
+        return all;
+    }
+
+    Map<String, DatabaseReaderLazyLoader> getDefaultDatabases() {
+        return defaultDatabases;
+    }
+
+    Map<String, DatabaseReaderLazyLoader> getConfigDatabases() {
+        return configDatabases;
+    }
+
+    void updateDatabase(Path file, boolean update) {
+        String databaseFileName = file.getFileName().toString();
+        try {
+            Map<String, DatabaseReaderLazyLoader> databases = new HashMap<>(this.configDatabases);
```

Comment:
I think we should use `ConcurrentHashMap` for `this.configDatabases`, then we don't have to do copying and it should be safer for multi-threading. 
Currently there's no guarantee that `this.configDatabases` in line 109 is the same as in 127 - get-and-update access should always include while loop and atomic swap (or something like `Atomic<Something>.updateAndGet()`)

Reply:
So I picked this approach because of the following reason:
- I think this method will not be invoked concurrently if there are multiple file updates. At least this is how understand the `ResourceFileWatcher`. The synchronizations happens here: `ResourceWatcherService` line 165, the `run()` method is synchronized. This then delegates to `FileWatcher.FileObserver#checkAndNotify()` and checks whether the files to be monitored have changed and then invoke this `GeoipDirectoryListener` for each change.
- I opted for the copy on write update mechanism here, because that is more efficient in cases where incidental changes are made and the resource is mainly read. This is a pattern we use elsewhere in the code base, where there is a resources, that is mainly read and rarely updated.

I'm okay with changing the configDatabases field to use `ConcurrentHashMap`, then we don't need the update logic here anymore, which makes this class simpler.

Reply:
I'd argue that `ConcurrentHashMap` will be actually faster here - `get` has no penalty at all and `put`/`remove` has to do less than copying whole map (with no contention it's just 1)find slot 2)add to list, done). And it gives you safety if we ever refactor `updateDatabase` to be invoked concurrently

Reply:
👍 let's then go with `ConcurrentHashMap `